### PR TITLE
Text on proof automation

### LINKED ISSUE #920: WRITING: Make a first draft of the "Automated Theorem Proving" section of the paper
See the discussion at https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/What.20are.20the.20hardest.20positive.20implications.20for.20an.20ATP.3F

### PR DESCRIPTION
**Pull request by goens**
See #920 

The current structure (and even title modification) are just suggestions, happy to discuss them in this PR of course!

Closes #920

### CODE REVIEW

File: paper/automated.tex [hunk 1]
```diff
@@ -1,20 +1,108 @@
-\section{Automated Theorem Proving}\label{automated-sec}
+\section{Proof Automation}\label{automated-sec}
 
-\note{TODO: expand this sketch}
+In this project we used proof automation in two ways: automated theorem provers (ATPs) and Lean tactics.
+ATPs are generally stand-alone tools that implement a (semi-) decision procedure for a given formal language or related set of languages.
+For example, Vampire~\cite{DBLP:conf/cav/KovacsV13} is an ATP focused primarily on first-order logic using supperposition, which we used extensively in this project.
```
Comment (vlad902) [inline paper/automated.tex:5]: supper->super

File: paper/automated.tex [hunk 2]
```diff
@@ -1,20 +1,108 @@
...
+ATPs are generally complex pieces of software, and can often have bugs.
+While we could choose to trust the results of an ATP, and in this project we don't do so either.
```
Comment (vlad902) [inline paper/automated.tex:7]: Needs rewrite.

File: paper/automated.tex [hunk 3]
```diff
@@ -1,20 +1,108 @@
...
+Instead, we use proof certificates, which many ATPs can produce, to reconstruct a proof in Lean.
+This process depends on the proof certificate and the ATP, and we describe it for the main recunstruction we've done.
```
Comment (vlad902) [inline paper/automated.tex:9]: recunstruction -> reconstruction. we've -> we have.

File: paper/automated.tex [hunk 4]
```diff
@@ -1,20 +1,108 @@
...
 
-Describe various automated theorem provers deployed in the project, with some statistics on performance:
+Tactics in Lean, on the other hand, are meta-programs~\cite{DBLP:journals/pacmpl/EbnerURAM17} that builds proofs.
+In other words, tactics are programs that operate at the meta-level of Lean code: they essentially take in Lean code as input and produce Lean code as output.
+In this manner, they look like another keyword in the language, and are tightly integrated by producing proofs directly.
+Under the hood, they can implement essentially anything, from syntactic sugar to full decision procedures.
+The \texttt{duper} tactic~\cite{DBLP:conf/itp/CluneQBA24}, for example, implements a superposition calculus, similar Vampire's, but for dependent types --- Lean's underlying logical foundation.
```
Comment (vlad902) [inline paper/automated.tex:15]: similar to

File: paper/automated.tex [hunk 5]
```diff
@@ -1,20 +1,108 @@
...
+The main two families of ATPs and tactics we used here are superposition/saturation-based and equational reasoning ones.
+In this context we also include SMT solvers, which combine specific decision procedures for theories, like congruence closure for equational reasoning, with satisfiability (SAT) solving~\cite{}.
+Finally, we also used \texttt{aesop}~\cite{DBLP:conf/cpp/LimpergF23}, which implements a version of tableau search.
+This was used mainly to help specific constructions in refutations, and is not specific to proving or disproving magma implications in this sense.
+We will describe our use of \texttt{aesop} more in Section~\ref{sec:proof-reconstruction} below.
+
+\paragraph{Saturation}
+Most of the ATPs we used extensively in this project rely primarily on saturation procdures in the superposition calculus.
+This is the case for Vampire~\cite{DBLP:conf/cav/KovacsV13}, and the paper alo gives a more in-depth explanation of this.
```
Comment (vlad902) [inline paper/automated.tex:30]: alo -> also. though i'd probably rephrase this to use the citation as the object in that clause.

File: paper/automated.tex [hunk 6]
```diff
@@ -1,20 +1,108 @@
...
+See also~\cite{DBLP:journals/cacm/BentkampBNTVW23} for a gentler exposition.
+The core idea of these provers is that they take a set of assumptions and conjecture, expressed in --- say --- first-order logic.
+They take the conjecture and negate it, adding this negation to the set of assumptions, which are all put in some normal form.
+The ATP then try to refute the negation by applying rules of the underlying calculus, until they find a proof of false (a contradiction).
```
Comment (vlad902) [inline paper/automated.tex:34]: try -> tries

File: paper/automated.tex [hunk 7]
```diff
@@ -1,20 +1,108 @@
...
+In this case, the conjecture was (classicaly) true, and the ATP has found a proof by contradiction, often called a ``refutation'' or ``saturation'' proof.
+
+The underlying calculus varies from system to system, but they often have a variant of a resolution clause, a clause of a form:
+\[\infer{C \lor D}{C \lor L \quad D \lor \neg L} \]
+This can also be read as $C \lor L \quad D \lor \neg L$ implies $C \lor D$, where $C, D, L$ are formulas in e.g. first-order logic.
+Superposition calculi have a variant of this rule that deals with equality directly, and thus are more efficient at reasoning about equality.
+
+In this project we used Vampire~\cite{DBLP:conf/cav/KovacsV13}, Duper~\cite{DBLP:conf/itp/CluneQBA24} and Prover9 and Mace4~\cite{prover9-mace4} which are all based on variants of saturation for proving.
```
Comment (vlad902) [inline paper/automated.tex:42]: Does mace4 count as a saturation-based prover? I'm not an expert here.

File: paper/automated.tex [hunk 8]
```diff
@@ -1,20 +1,108 @@
...
+On the other hand, equality saturation~\cite{DBLP:journals/pacmpl/WillseyNWFTP21} uses e-graphs by extending congruence closure to a more controlled search, enabling optimization and conditional rewriting.
+One of the main advantages of using equational reasoning to reason about implications of magma laws is that we get very explicit proofs: a proof that $l \vdash l'$ is given by a sequence of rewrites that starts at the left-hand side of $l'$ and arrives at the right-hand side through applications of $l$.
+
+In this project we used Z3~\cite{DBLP:conf/tacas/MouraB08}, Prover9 and Mace4~\cite{prover9-mace4}, a custom ATP for magmas based on egg~\cite{DBLP:journals/pacmpl/WillseyNWFTP21}, and the Lean egg tactic~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24,rossel2024bridging}, which all work with equational logic. We have also reasoned with manual (custom written) heuristics about simple rewrites.
+
+\subsection{Proof Reconstruction and Integration}
+\label{sec:proof-reconstruction}
+
+While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean (except for Lean tactics like \texttt{duper}, \texttt{aesop} and \texttt{egg}).
```
Comment (vlad902) [inline paper/automated.tex:60]: "While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean" I would be more specific: "ATPs can resolve many implications or non-implications studied by this project; however, except for Lean tactics like ..., ATP systems do not generate proof certificates that are natively recognized by Lean."

File: paper/automated.tex [hunk 9]
```diff
@@ -1,20 +1,108 @@
...
+    \item Using a Lean tactic, which calls a decision procedure written in Lean (like \texttt{aesop} or \texttt{duper}).
+    \item Using a Lean tactic, which calls an existing (external) ATP and reconstructs a proof term from the ATP's result (like \texttt{bv\_decide} or \texttt{egg}).
+    \item\label{external} Using an external script which calls an existing ATP and generates a source file \texttt{.lean} which captures the result explicitly.
+\end{enumerate}
+
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
```
Comment (vlad902) [inline paper/automated.tex:83]: become -> becomes. I don't understand what this is conveying

File: paper/automated.tex [hunk 10]
```diff
@@ -1,20 +1,108 @@
...
+\begin{enumerate}
+    \item Using a Lean tactic, which calls a decision procedure written in Lean (like \texttt{aesop} or \texttt{duper}).
+    \item Using a Lean tactic, which calls an existing (external) ATP and reconstructs a proof term from the ATP's result (like \texttt{bv\_decide} or \texttt{egg}).
+    \item\label{external} Using an external script which calls an existing ATP and generates a source file \texttt{.lean} which captures the result explicitly.
+\end{enumerate}
+
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
```
Comment (vlad902) [inline paper/automated.tex:82]: Magmas -> magmas

File: paper/automated.tex [hunk 11]
```diff
@@ -1,20 +1,108 @@
...
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
+
+\paragraph{Semi-Automated Counterexample Guidance}
+
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
+The general strategy was to use ATPs to find counter-examples to implications by building magmas iteratively.
+If we want to build a counterexample to $l \vdash l'$, we want to build construct a magma where $l$ holds but $l'$ doesn't.
+In this method, we iteratively strengthen a construction with additional hypotheses, and use the ATP to check wether these hypotheses are not too strong (to imply $l'$) or unsound (to disallow $l$).
```
Comment (vlad902) [inline paper/automated.tex:90]: wether -> whether

File: paper/automated.tex [hunk 12]
```diff
@@ -1,20 +1,108 @@
...
+
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
+
+\paragraph{Semi-Automated Counterexample Guidance}
+
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
```
Comment (vlad902) [inline paper/automated.tex:87]: I would maybe get rid of "in a semi-automatic fashion," here. There are examples of fully-automated and quite bespoke counterexample construction. It may be interesting to mention that, or leave a TODO for the future.

File: paper/automated.tex [hunk 13]
```diff
@@ -1,20 +1,108 @@
...
+The general strategy was to use ATPs to find counter-examples to implications by building magmas iteratively.
+If we want to build a counterexample to $l \vdash l'$, we want to build construct a magma where $l$ holds but $l'$ doesn't.
+In this method, we iteratively strengthen a construction with additional hypotheses, and use the ATP to check wether these hypotheses are not too strong (to imply $l'$) or unsound (to disallow $l$).
+
+TODO: this should also be expanded more, at least with references to some of the constructions in other chapters.
+
+While equational theories can also be used in a semi-automatic fashion to prove equations~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24}, the positive implications in this project were all simple enough that we did not need a semi-automatic approach for them.
```
Comment (vlad902) [inline paper/automated.tex:94]: equational theories -> equality saturation? While this sentence is true for the general graph, it is not completely true for either the finite implications or the Higman-Neumann work Jose Brox has done, where he has indeed performed guided search.

File: paper/automated.tex (hunk 2 above)
Comment (vlad902) [reply paper/automated.tex:7]: I would merge with the previous/next sentence and make it more explicit: "ATPs are complex software that can contain bugs. Instead of trusting ATP output, we used proof certificates, which many ATPs can produce, to reconstruct a proof in Lean.

File: paper/automated.tex [hunk 14]
```diff
@@ -1,20 +1,108 @@
...
+We will describe our use of \texttt{aesop} more in Section~\ref{sec:proof-reconstruction} below.
+
+\paragraph{Saturation}
+Most of the ATPs we used extensively in this project rely primarily on saturation procdures in the superposition calculus.
+This is the case for Vampire~\cite{DBLP:conf/cav/KovacsV13}, and the paper alo gives a more in-depth explanation of this.
+See also~\cite{DBLP:journals/cacm/BentkampBNTVW23} for a gentler exposition.
+The core idea of these provers is that they take a set of assumptions and conjecture, expressed in --- say --- first-order logic.
+They take the conjecture and negate it, adding this negation to the set of assumptions, which are all put in some normal form.
+The ATP then try to refute the negation by applying rules of the underlying calculus, until they find a proof of false (a contradiction).
+In this case, the conjecture was (classicaly) true, and the ATP has found a proof by contradiction, often called a ``refutation'' or ``saturation'' proof.
```
Comment (teorth) [inline paper/automated.tex:35]: "classicaly" -> "classically"

File: paper/automated.tex [hunk 15]
```diff
@@ -1,20 +1,108 @@
...
+
+SMT solvers like Z3~\cite{DBLP:conf/tacas/MouraB08} use equational reasoning for deciding the theory of equality with uninterpreted functions~\cite{DBLP:series/txtcs/KroeningS16,DBLP:conf/cade/MouraB07}.
+On the other hand, equality saturation~\cite{DBLP:journals/pacmpl/WillseyNWFTP21} uses e-graphs by extending congruence closure to a more controlled search, enabling optimization and conditional rewriting.
+One of the main advantages of using equational reasoning to reason about implications of magma laws is that we get very explicit proofs: a proof that $l \vdash l'$ is given by a sequence of rewrites that starts at the left-hand side of $l'$ and arrives at the right-hand side through applications of $l$.
+
+In this project we used Z3~\cite{DBLP:conf/tacas/MouraB08}, Prover9 and Mace4~\cite{prover9-mace4}, a custom ATP for magmas based on egg~\cite{DBLP:journals/pacmpl/WillseyNWFTP21}, and the Lean egg tactic~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24,rossel2024bridging}, which all work with equational logic. We have also reasoned with manual (custom written) heuristics about simple rewrites.
+
```
Comment (teorth) [inline paper/automated.tex:56]: It would be great to have some illustrations (perhaps as figures or screenshots) of some sample outputs (e.g. a command line interface input and output) of some of these ATP tools in action for a sample implication in the project.  Perhaps one can add one such example somewhere around here and make a stub to ask for other contributions of this type?  Ideally one should have examples from two or three different representative ATPs.

File: paper/automated.tex [hunk 16]
```diff
@@ -1,20 +1,108 @@
...
+These two aspect present different challenges, and require different strategies, depending mostly on the kind of proof strategy the decision procedure uses.
+
+For saturation proofs $\ldots$ (TODO: add explanations from \url{https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/Vampire.20-.3E.20Lean/near/478147138})
+
+For equational proofs from external provers (e.g. MagmaEgg), we used also a simple version of reconstruction by (re-)constructing the proofs of equality from an explanation, using congruence lemmas from Lean.
+In the case of equational proofs from the \texttt{egg} tactic, these could be converted into a series of \texttt{calc} steps, documenting the explicit calculation, using the \texttt{calcify} tactic\footnote{\url{https://github.com/nomeata/lean-calcify/}}.
+
+In general, we've observed that there are multiple ways of integrating decisions procedures within Lean, with different levels of integration.
```
Comment (teorth) [inline paper/automated.tex:72]: I would generally avoid contractions such as "we've" in academic writing.

File: paper/automated.tex (hunk 1 above)
Comment (pitmonticone) [reply paper/automated.tex:5]: Fixed. Thanks.

File: paper/automated.tex (hunk 3 above)
Comment (pitmonticone) [reply paper/automated.tex]: Fixed. Thanks.

File: paper/automated.tex (hunk 4 above)
Comment (pitmonticone) [reply paper/automated.tex:15]: Fixed. Thanks.

File: paper/automated.tex (hunk 6 above)
Comment (pitmonticone) [reply paper/automated.tex:34]: Fixed. Thanks.

File: paper/automated.tex (hunk 10 above)
Comment (pitmonticone) [reply paper/automated.tex:82]: Fixed. Thanks.

File: paper/automated.tex (hunk 11 above)
Comment (pitmonticone) [reply paper/automated.tex:90]: Fixed. Thanks.

File: paper/automated.tex (hunk 9 above)
Comment (pitmonticone) [reply paper/automated.tex]: Fix the typo, but I will not resolve the conversation because of the comment.

**Comment by pitmonticone** [comment]
I have directly fixed a bunch of typos I have found together with all those found by @vlad902 and resolved the relevant review comments.

File: paper/automated.tex (hunk 9 above)
Comment (goens) [reply paper/automated.tex]: I don't really understand what you don't understand about it. I expanded it more with a citation, hopefully that helps.

File: paper/automated.tex (hunk 12 above)
Comment (goens) [reply paper/automated.tex:87]: I don't understand the issue with that, but I have also been less involved in this effort. Semi-automated subsumes fully automated, and not all have been fully automated, right? More importantly though, the entire section has talked about fully automated approaches, this is just saying it's not *only* been fully automated, there have been semi-automated approaches. Maybe this should be expanded in a different PR after?

File: paper/automated.tex (hunk 13 above)
Comment (goens) [reply paper/automated.tex:94]: fair point! I changed the formulation a bit and added a TODO to include those.

**Comment by goens** [comment]
I think I've addressed all the suggestions now. Happy to have either a second round of reviews/suggestions or merge this and continue expanding with more PRs.

**Comment by teorth** [comment]
I think it's good to merge for now.  If anyone wants to work on the section more, they can say so in the Zulip chat, and I'll open a new issue for it.

File: paper/automated.tex (hunk 9 above)
Comment (vlad902) [reply paper/automated.tex]: After re-reading I'm not sure what my original confusion was about.
